### PR TITLE
Add command to verify presence of replays on S3

### PR DESCRIPTION
**Pull request by peppy**
A simple one. To be used hopefully once-ever, to fix https://github.com/orgs/ppy/projects/19/views/1?pane=issue&itemId=51932194.

https://github.com/orgs/ppy/projects/19/views/1?pane=issue&itemId=51932194

### CODE REVIEW

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+            var s3Key = Environment.GetEnvironmentVariable("S3_KEY") ?? string.Empty;
+            var s3Secret = Environment.GetEnvironmentVariable("S3_SECRET") ?? string.Empty;
+
+            using var s3 = new AmazonS3Client(new BasicAWSCredentials(s3Key, s3Secret), new AmazonS3Config
+            {
+                CacheHttpClient = true,
+                HttpClientCacheSize = 32,
+                RegionEndpoint = RegionEndpoint.USWest1,
+                UseHttp = true,
```
Comment (bdach) [inline osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:53]: http (as opposed to https, I imagine, although the docs are [crap](https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/Runtime/TClientConfig.html))? Sure about that?

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+                if (!importedScores.Any())
+                {
+                    if (lastId > await conn.QuerySingleAsync<ulong>("SELECT MAX(id) FROM scores"))
+                    {
+                        Console.WriteLine("All done!");
+                        break;
+                    }
+
+                    lastId += (ulong)BatchSize;
```
Comment (bdach) [inline osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:83]: Why is `lastId` incremented only in this branch? Doesn't this mean that this is going to spin over the first batch of scores forevermore?

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            Console.WriteLine();
+            Console.WriteLine($"Verifying replay flag for scores starting from {lastId}");
+
+            if (DryRun)
+                Console.WriteLine("RUNNING IN DRY RUN MODE.");
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                IEnumerable<SoloScore> importedScores = await conn.QueryAsync<SoloScore>(
+                    "SELECT `id`, `has_replay` FROM scores WHERE id BETWEEN @lastId AND (@lastId + @batchSize - 1) AND has_replay = 1 AND legacy_score_id IS NULL ORDER BY id",
```
Comment (bdach) [inline osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:68]: Why select `has_replay` if the filter already specifies `has_replay = 1`?

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                        break;
+                    }
+
+                    lastId += (ulong)BatchSize;
+                    continue;
+                }
+
+                foreach (var score in importedScores)
+                {
+                    if (!score.has_replay) continue;
```
Comment (bdach) [inline osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:89]: In conjunction with the above, why is this being checked at all? `has_replay` should always be true making this a dead branch.

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+                    try
+                    {
+                        // check the replay is available on s3.
+                        _ = await s3.GetObjectMetadataAsync("score-replays", score.id.ToString(CultureInfo.InvariantCulture), cancellationToken);
```
Comment (bdach) [inline osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:94]: As already touched on above, this feels quite dodge. Any old network failure and this deletes replays. No error logging either. Is there really no better way to at least check for a slightly more specific error?

There's [some old code online that does this](https://github.com/aws/aws-sdk-net/blob/c688adc3850fbcbf708154258963afbb680b59d1/sdk/src/Services/S3/Custom/_bcl/IO/S3FileInfo.cs#L104-L147), but then again, not sure how up-to-date that is, given `S3FileInfo.Exists()` was apparently a method offered out of the box in AWS SDK, but now it doesn't seem to be there at all anymore...

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs (hunk 1 above)
Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:53]: This was just copypaste from `osu-server-spectator`. I believe there was probably a reason for this (cloudflare r2 support maybe?) at some point.

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs (hunk 5 above)
Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:94]: This was also just copypaste from our S3 helper utility in `osu-server-spectator`. But I agree it's pretty dodge.

Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:93]: I'll give what you linked a try, but the exception being thown is `PermissionDenied`, which is the same for a multitude of cases including `NotFound`. I'm not sure we get any more fidelity in exceptions these days from S3.

Comment (bdach) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs:93]: Sure, I'd still say matching on that is better than indiscriminate catching.

**Review by bdach — APPROVED** [review]
Seems ok

**Pull request merged**